pkg/controller/volume/cache/actual_state_of_world.go

```diff
@@ -190,63 +200,78 @@ func (asw *actualStateOfWorld) AddVolumeNode(volumeSpec *volume.Spec, nodeName s
 		volumeObj.nodesAttachedTo[nodeName] = nodeAttachedTo{
-			nodeName:            nodeName,
-			safeToDetach:        false,
-			detachRequestedTime: time.Time{},
+			nodeName:              nodeName,
+			mountedByNode:         true, // Assume mounted, until proven otherwise
+			mountedByNodeSetCount: 0,
+			detachRequestedTime:   time.Time{},
 		}
+	} else if !nodeObj.detachRequestedTime.IsZero() {
+		// Reset detachRequestedTime values if object exists and time is non-zero
+		nodeObj.detachRequestedTime = time.Time{}
+		volumeObj.nodesAttachedTo[nodeName] = nodeObj
 	}
 
 	return volumeName, nil
 }
 
-func (asw *actualStateOfWorld) MarkVolumeNodeSafeToDetach(
-	volumeName, nodeName string) error {
+func (asw *actualStateOfWorld) SetVolumeMountedByNode(
+	volumeName api.UniqueDeviceName, nodeName string, mounted bool) error {
 	asw.Lock()
 	defer asw.Unlock()
 	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
 	if !volumeExists {
 		return fmt.Errorf(
-			"failed to MarkVolumeNodeSafeToDetach(volumeName=%q, nodeName=%q) volumeName does not exist",
+			"failed to SetVolumeMountedByNode(volumeName=%v, nodeName=%q, mounted=%v) volumeName does not exist",
 			volumeName,
-			nodeName)
+			nodeName,
+			mounted)
 	}
 
 	nodeObj, nodeExists := volumeObj.nodesAttachedTo[nodeName]
 	if !nodeExists {
 		return fmt.Errorf(
-			"failed to MarkVolumeNodeSafeToDetach(volumeName=%q, nodeName=%q) nodeName does not exist",
+			"failed to SetVolumeMountedByNode(volumeName=%v, nodeName=%q, mounted=%v) nodeName does not exist",
 			volumeName,
-			nodeName)
+			nodeName,
+			mounted)
+	}
+
+	if mounted {
+		// Increment set count
+		nodeObj.mountedByNodeSetCount = nodeObj.mountedByNodeSetCount + 1
+	} else {
+		// Do not allow value to be reset unless it has been set at least once
+		if nodeObj.mountedByNodeSetCount == 0 {
+			return nil
+		}
 	}
 
-	// Reset safe to detach
-	nodeObj.safeToDetach = true
+	nodeObj.mountedByNode = mounted
 	volumeObj.nodesAttachedTo[nodeName] = nodeObj
 
 	return nil
 }
 
 func (asw *actualStateOfWorld) MarkDesireToDetach(
-	volumeName, nodeName string) (time.Duration, error) {
+	volumeName api.UniqueDeviceName, nodeName string) (time.Duration, error) {
 	asw.Lock()
 	defer asw.Unlock()
 
 	volumeObj, volumeExists := asw.attachedVolumes[volumeName]
 	if !volumeExists {
 		return time.Millisecond * 0, fmt.Errorf(
-			"failed to MarkVolumeNodeSafeToDetach(volumeName=%q, nodeName=%q) volumeName does not exist",
+			"failed to MarkDesireToDetach(volumeName=%v, nodeName=%q) volumeName does not exist",
 			volumeName,
 			nodeName)
 	}
 
 	nodeObj, nodeExists := volumeObj.nodesAttachedTo[nodeName]
 	if !nodeExists {
 		return time.Millisecond * 0, fmt.Errorf(
-			"failed to MarkVolumeNodeSafeToDetach(volumeName=%q, nodeName=%q) nodeName does not exist",
+			"failed to MarkDesireToDetach(volumeName=%v, nodeName=%q) nodeName does not exist",
 			volumeName,
 			nodeName)
 	}
```

Comment:
Discussed offline. Object is created with time zero. `MarkDesireToDetach` sets it to current time, if zero, and returns time elapsed between non-zero time and `Now`. `AddVolumeNode` resets the time to zero.